File: ClassicLib/VersionRegistry/matching.py
"""Version matching algorithms with graceful degradation.

This module provides intelligent matching of detected game versions to known
versions in the registry, with configurable fallback strategies for unknown
versions.

The matching algorithm follows this priority order:
1. Exact match - Version string matches exactly
2. Range match - Version falls within a defined compatible_range
3. Nearest match - Same major version, closest minor/patch
4. Default fallback - Use highest priority version for the game

Example:
    >>> from ClassicLib.VersionRegistry.matching import VersionMatcher, MatchConfidence
    >>> from ClassicLib.VersionRegistry import get_version_registry
    >>> registry = get_version_registry()
    >>> matcher = VersionMatcher(registry)
    >>> result = matcher.match(Version("1.10.163.0"), "Fallout4", is_vr=False)
    >>> result.confidence
    <MatchConfidence.EXACT: 1>

"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from typing import TYPE_CHECKING

from ClassicLib.Logger import logger

if TYPE_CHECKING:
    from collections.abc import Sequence

    from packaging.version import Version

    from ClassicLib.VersionRegistry.core import VersionRegistry
    from ClassicLib.VersionRegistry.models import VersionInfo


class MatchConfidence(Enum):
    """Confidence level for version match.

    Indicates how confident the matching algorithm is about the result.
    Higher confidence means a more reliable match.

    Attributes:
        EXACT: Exact version match - highest confidence.
        RANGE: Version falls within a defined compatible range.
        NEAREST: Matched to nearest known version by distance calculation.
        DEFAULT: Using default fallback for the game.
        UNKNOWN: No match found at all.

    """

    EXACT = auto()
    RANGE = auto()
    NEAREST = auto()
    DEFAULT = auto()
    UNKNOWN = auto()


@dataclass
class MatchResult:
    """Result of version matching.

    Contains the matched version information along with metadata about
    the match quality and any relevant messages for the user.

    Attributes:
        version_info: Matched VersionInfo (may be None if UNKNOWN).
        confidence: How confident the match is.
        detected: Original detected version.
        message: Human-readable explanation of the match.

    Example:
        >>> result = MatchResult(
        ...     version_info=some_version_info,
        ...     confidence=MatchConfidence.EXACT,
        ...     detected=Version("1.10.163.0"),
        ...     message="Exact match: Fallout 4 Original"
        ... )
        >>> result.is_exact
        True
        >>> result.should_warn
        False

    """

    version_info: VersionInfo | None
    confidence: MatchConfidence
    detected: Version
    message: str = ""

    @property
    def is_exact(self) -> bool:
        """Check if this was an exact match.

        Returns:
            True if the match confidence is EXACT, False otherwise.

        """
        return self.confidence == MatchConfidence.EXACT

    @property
    def is_fallback(self) -> bool:
        """Check if this was a fallback match.

        A fallback match is one where the detected version was not found
        directly and a best-effort match was made.

        Returns:
            True if this was a fallback match (NEAREST, DEFAULT, or UNKNOWN).

        """
        return self.confidence in {
            MatchConfidence.NEAREST,
            MatchConfidence.DEFAULT,
            MatchConfidence.UNKNOWN,
        }

    @property
    def should_warn(self) -> bool:
        """Check if user should be warned about this match.

        Warnings are appropriate when the match is not exact or within
        a defined range, indicating potential compatibility issues.

        Returns:
            True if the user should be warned about this match.

        """
        return self.confidence in {
            MatchConfidence.NEAREST,
            MatchConfidence.DEFAULT,
        }

    @property
    def is_valid(self) -> bool:
        """Check if a valid version was matched.

        Returns:
            True if version_info is not None and confidence is not UNKNOWN.

        """
        return self.version_info is not None and self.confidence != MatchConfidence.UNKNOWN


class VersionMatcher:
    """Intelligent version matching with multiple fallback strategies.

    The matcher tries to find the best matching version for a detected
    game version, using the following order of preference:
    1. Exact version match
    2. Compatible range match
    3. Nearest major.minor version
    4. Default fallback for game

    Attributes:
        registry: The VersionRegistry instance to match against.

    Example:
        >>> from ClassicLib.VersionRegistry import get_version_registry
        >>> registry = get_version_registry()
        >>> matcher = VersionMatcher(registry)
        >>> result = matcher.match(Version("1.10.500.0"), "Fallout4", False)
        >>> result.confidence
        <MatchConfidence.NEAREST: 3>

    """

    def __init__(self, registry: VersionRegistry) -> None:
        """Initialize the VersionMatcher.

        Args:
            registry: The VersionRegistry instance to use for matching.

        """
        self._registry: VersionRegistry = registry

    def match(
        self,
        detected: Version,
        game: str = "Fallout4",
        is_vr: bool = False,
    ) -> MatchResult:
        """Match detected version to a known version in the registry.

        Attempts to find the best match for the detected version using
        multiple strategies in order of preference.

        Args:
            detected: The detected game version to match.
            game: Game identifier (e.g., "Fallout4").
            is_vr: Whether VR mode is active.

        Returns:
            MatchResult containing the matched version and confidence level.

        """
        # 1. Try exact match
        exact = self._registry.get_by_version(detected)
        if exact and exact.game == game and exact.is_vr == is_vr:
            return MatchResult(
                version_info=exact,
                confidence=MatchConfidence.EXACT,
                detected=detected,
                message=f"Exact match: {exact.display_name}",
            )

        # 2. Try compatible range match
        candidates = self._registry.get_all_for_game(game, is_vr)
        for candidate in candidates:
            if candidate.is_compatible_with(detected):
                return MatchResult(
                    version_info=candidate,
                    confidence=MatchConfidence.RANGE,
                    detected=detected,
                    message=f"Range match: {candidate.display_name} (compatible with {detected})",
                )

        # 3. Try nearest major.minor match
        nearest = self._find_nearest(detected, candidates)
        if nearest:
            logger.warning(f"Unknown version {detected} matched to nearest: {nearest.display_name}")
            return MatchResult(
                version_info=nearest,
                confidence=MatchConfidence.NEAREST,
                detected=detected,
                message=f"Nearest match: {nearest.display_name} (detected: {detected})",
            )

        # 4. Use default fallback
        if candidates:
            default = candidates[0]  # Highest priority (sorted by priority desc)
            logger.warning(f"Unknown version {detected} using default: {default.display_name}")
            return MatchResult(
                version_info=default,
                confidence=MatchConfidence.DEFAULT,
                detected=detected,
                message=f"Default fallback: {default.display_name} (detected: {detected})",
            )

        # 5. No match found
        logger.error(f"No version match found for {detected}")
        return MatchResult(
            version_info=None,
            confidence=MatchConfidence.UNKNOWN,
            detected=detected,
            message=f"No matching version found for {detected}",
        )

    def _find_nearest(  # noqa: PLR6301
        self,
        detected: Version,
        candidates: Sequence[VersionInfo],
    ) -> VersionInfo | None:
        """Find nearest version by semantic distance calculation.

        The distance is calculated as:
        - Major version difference * 1,000,000
        - Minor version difference * 1,000
        - Patch version difference * 1

        This ensures major version differences are heavily weighted.

        Args:
            detected: The detected version to find a match for.
            candidates: List of candidate versions to search.

        Returns:
            The nearest matching VersionInfo, or None if no reasonable match.

        """
        if not candidates:
            return None

        def distance(v: Version) -> float:
            """Calculate semantic version distance."""
            d_major = abs(detected.major - v.major) * 1_000_000
            d_minor = abs(detected.minor - v.minor) * 1_000
            d_patch = abs(detected.micro - v.micro)
            return d_major + d_minor + d_patch

        # Sort by distance, then by priority (higher priority = lower sort key)
        sorted_candidates = sorted(
            candidates,
            key=lambda c: (distance(c.version), -c.priority),
        )

        # Only return if reasonably close (same major version)
        best = sorted_candidates[0]
        if best.version.major == detected.major:
            return best

        return None

File: business-logic/classic-version-registry-core/src/matching.rs
//! Version matching algorithm.
//!
//! This module provides the `VersionMatcher` and `MatchResult` types for
//! intelligently matching unknown game versions to the nearest known version.

use crate::GameVersion;
use crate::models::VersionInfo;
use crate::registry::VersionRegistry;

/// Confidence level for version matching results.
///
/// Indicates how confident we are that the matched version is correct.
/// Lower confidence levels may warrant user warnings.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum MatchConfidence {
    /// Exact version match found in registry.
    Exact,
    /// Version falls within a defined compatible_range.
    Range,
    /// Matched to nearest known version by semantic distance.
    Nearest,
    /// Using default fallback version for the game.
    Default,
    /// No suitable match found.
    Unknown,
}

impl MatchConfidence {
    /// Check if this is a high-confidence match.
    #[must_use]
    pub const fn is_high_confidence(&self) -> bool {
        matches!(self, Self::Exact | Self::Range)
    }
}

/// Result of version matching.
///
/// Contains the matched version (if any), the confidence level,
/// and a message describing the match result.
#[derive(Debug, Clone)]
pub struct MatchResult {
    /// The matched version info, if found.
    pub version_info: Option<VersionInfo>,
    /// Confidence level of the match.
    pub confidence: MatchConfidence,
    /// The originally detected version.
    pub detected: GameVersion,
    /// Human-readable message about the match.
    pub message: String,
}

impl MatchResult {
    /// Create a new match result.
    #[must_use]
    pub fn new(
        version_info: Option<VersionInfo>,
        confidence: MatchConfidence,
        detected: GameVersion,
        message: impl Into<String>,
    ) -> Self {
        Self {
            version_info,
            confidence,
            detected,
            message: message.into(),
        }
    }

    /// Check if this is an exact match.
    #[must_use]
    pub const fn is_exact(&self) -> bool {
        matches!(self.confidence, MatchConfidence::Exact)
    }

    /// Check if this is a fallback match (Nearest, Default, or Unknown).
    #[must_use]
    pub const fn is_fallback(&self) -> bool {
        matches!(
            self.confidence,
            MatchConfidence::Nearest | MatchConfidence::Default | MatchConfidence::Unknown
        )
    }

    /// Check if the user should be warned about this match.
    ///
    /// Returns `true` for Nearest and Default matches, where the
    /// matched version may not be exactly what the user has.
    #[must_use]
    pub const fn should_warn(&self) -> bool {
        matches!(
            self.confidence,
            MatchConfidence::Nearest | MatchConfidence::Default
        )
    }

    /// Check if this is a valid match (version_info is present and not Unknown).
    #[must_use]
    pub fn is_valid(&self) -> bool {
        self.version_info.is_some() && self.confidence != MatchConfidence::Unknown
    }
}

/// Version matcher for finding the best matching version.
///
/// Implements the matching algorithm with the following priority:
/// 1. Exact match
/// 2. Compatible range match
/// 3. Nearest match (same major version)
/// 4. Default fallback
/// 5. Unknown
pub struct VersionMatcher<'a> {
    registry: &'a VersionRegistry,
}

impl<'a> VersionMatcher<'a> {
    /// Create a new version matcher.
    #[must_use]
    pub const fn new(registry: &'a VersionRegistry) -> Self {
        Self { registry }
    }

    /// Match a detected version to the registry.
    ///
    /// # Arguments
    ///
    /// * `detected` - The detected game version
    /// * `game` - Game identifier (e.g., "Fallout4")
    /// * `is_vr` - Whether VR mode is active
    ///
    /// # Returns
    ///
    /// A `MatchResult` with the best matching version and confidence level.
    pub fn match_version(&self, detected: &GameVersion, game: &str, is_vr: bool) -> MatchResult {
        // Try exact match first
        if let Some(result) = self.find_exact(detected, game, is_vr) {
            return result;
        }

        // Try range match
        if let Some(result) = self.find_range(detected, game, is_vr) {
            return result;
        }

        // Try nearest match
        if let Some(result) = self.find_nearest(detected, game, is_vr) {
            return result;
        }

        // Try default fallback
        if let Some(result) = self.find_default(detected, game, is_vr) {
            return result;
        }

        // No match found
        MatchResult::new(
            None,
            MatchConfidence::Unknown,
            *detected,
            format!(
                "No matching version found for {} (game: {}, VR: {})",
                detected, game, is_vr
            ),
        )
    }

    /// Find an exact version match.
    fn find_exact(&self, detected: &GameVersion, game: &str, is_vr: bool) -> Option<MatchResult> {
        let version = self.registry.get_by_version(detected)?;

        // Verify game and VR mode match
        if version.game != game || version.is_vr != is_vr {
            return None;
        }

        Some(MatchResult::new(
            Some(version.clone()),
            MatchConfidence::Exact,
            *detected,
            format!("Exact match: {} ({})", version.display_name, version.id),
        ))
    }

    /// Find a version within a compatible range.
    fn find_range(&self, detected: &GameVersion, game: &str, is_vr: bool) -> Option<MatchResult> {
        for version in self.registry.get_all_for_game(game, Some(is_vr)) {
            if let Some(range) = &version.compatible_range
                && range.contains(detected)
            {
                return Some(MatchResult::new(
                    Some(version.clone()),
                    MatchConfidence::Range,
                    *detected,
                    format!(
                        "Version {} falls within compatible range for {}",
                        detected, version.display_name
                    ),
                ));
            }
        }
        None
    }

    /// Find the nearest version by semantic distance.
    ///
    /// Only matches versions with the same major version number.
    fn find_nearest(&self, detected: &GameVersion, game: &str, is_vr: bool) -> Option<MatchResult> {
        let candidates: Vec<_> = self
            .registry
            .get_all_for_game(game, Some(is_vr))
            .into_iter()
            .filter(|v| v.version.same_major(detected))
            .collect();

        if candidates.is_empty() {
            return None;
        }

        // Find the nearest by semantic distance
        let mut best = candidates[0];
        let mut best_distance = detected.semantic_distance(&best.version);

        for candidate in &candidates[1..] {
            let distance = detected.semantic_distance(&candidate.version);
            // If same distance, prefer higher priority
            if distance < best_distance
                || (distance == best_distance && candidate.priority > best.priority)
            {
                best = *candidate;
                best_distance = distance;
            }
        }

        Some(MatchResult::new(
            Some(best.clone()),
            MatchConfidence::Nearest,
            *detected,
            format!(
                "Version {} matched to nearest: {} (distance: {})",
                detected, best.display_name, best_distance
            ),
        ))
    }

    /// Find the default version for the game.
    fn find_default(&self, detected: &GameVersion, game: &str, is_vr: bool) -> Option<MatchResult> {
        // Get highest priority version for the game
        let versions = self.registry.get_all_for_game(game, Some(is_vr));

        if versions.is_empty() {
            return None;
        }

        // Already sorted by priority descending
        let default = versions[0];

        Some(MatchResult::new(
            Some(default.clone()),
            MatchConfidence::Default,
            *detected,
            format!(
                "Using default version {} for unrecognized version {}",
                default.display_name, detected
            ),
        ))
    }
}

#[cfg(test)]
#[path = "matching_tests.rs"]
mod tests;
